### ISSUE TITLE
Choose an editor for creating topics/posts.

### ISSUE THREAD

**Issue by euantorano** [mybb/mybb2#304]
We need to select a JS based WYSIWYG editor for 2.0.

This editor will be used when creating posts, conversations and other content.

**Requirements**

- MyCode/BBCode support
- Markdown support
- Ability to disable WYSIWYG feature, showing plain BBCode/Markdown and remember this choice (and allow this choice to be set via User CP/Admin CP)
- Ability to easily add new buttons/menus to the editor toolbar
- Open Source and permissively licensed (MIT, BSD3 or similar)
- Good display and performance on mobile

**Nice to have**

- Full screen edit mode

**Comment by Shade-** [comment]
How will data be stored and processed? If we're going to store already parsed data and an original content to allow edits, https://quilljs.com would be the nicest of them all IMHO. It needs a Quill-to-BBCode/Markdown parser, but its API is well documented.

**Comment by euantorano** [comment]
That's a good question. Currently we store data pre-parsed, but there are a few downsides:

- BBCode/MyCode when added or modified won't be supported until a post is edited
- `@` style mentions won't track username changes, leading to confusion

@dvz Suggested that for the second issue we can pre-parse content that doesn't contain mentions, and otherwise parse on display. This is a possibility.

I think that really we need to benchmark the parser to make a decision.

**Comment by AdrianKuriata** [comment]
I suggest summernote, big machine and very configurable.

**Comment by kawaii** [comment]
I'm not sure if this is somewhere we can suggest editors but my vote goes for Monaco:

https://microsoft.github.io/monaco-editor/

**Comment by euantorano** [comment]
Yes, this is as good a place as any to suggest editors. A good suggestion should explain the features provided, the license, etc.

**Comment by Broatcast** [comment]
Have a look on StackEdit
https://github.com/benweet/stackedit 
https://stackedit.io/

a nice and clean editore, i love the live preview function and its easy to customise to

**Comment by Shade-** [comment]
Regarding [this post and related ones](https://community.mybb.com/thread-214721-post-1292658.html#pid1292658), if we are really going the mixed parser way I guess Quill is the only choice we have. I've been looking for already-built editors with BBCode and Markdown support but either they do not support one of the above, or they support them separately.

The main advantage of Quill is that it uses its own logic and formatting, which can be then converted to whatever you like. That's why I would choose it, even if we'll need to write our own Quill-to-BBCode/Markdown JavaScript parser.

**Comment by effone** [comment]
Certainly not fulfilling all of our requirements, but have just come across this feature-rich Markdown editor and wanted to share with you guys:
https://github.com/nhnent/tui.editor

**Comment by effone** [comment]
Also, I guess this [MarkItUp!](http://markitup.jaysalvat.com/) a good start?

Its an old one, light-weight, very flexible to suit our need ... This is not a full fledged editor but an engine to make editor as per requirement on top of it, as they say ...
Supports HTML, BBCode, Textile, Markdown, Dotclear, Wiki even own custom markups << This may be helpful for custom MyCodes ?
https://github.com/markitup/1.x
